src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,164 @@
+     */
+    public static TaskList findKeyword(String str, TaskList tasks) throws InvalidKeywordException {
+        if (str.split(" ").length == 2) {
+            String keyword = str.split(" ")[1];
+            return tasks.findTask(keyword);
+        } else {
+            throw new InvalidKeywordException("Keyword given is not a single word.");
+        }
+    }
+
+    /**
+     * Extracts the task index from a user input string and deletes the task.
+     *
+     * @param str    The user input string.
+     * @param tasks  The list of tasks to operate on.
+     * @return The task that was deleted.
+     * @throws InvalidTaskIndexException   If the task index is invalid.
+     * @throws MissingTaskIndexException  If the task index is missing.
+     */
+    public static Task taskToDelete(String str, TaskList tasks)
+            throws InvalidTaskIndexException, MissingTaskIndexException{
+        if (str.split(" ").length == 2) {
+            int taskIndex = Integer.parseInt(str.split(" ")[1]) - 1;
+            if (taskIndex + 1 > tasks.getSize() || taskIndex < 0) {
+                throw new InvalidTaskIndexException("Invalid Task Index.");
+            }
+            Task toRemove = tasks.getTask(taskIndex);
+            tasks.deleteTask(taskIndex);
+            return toRemove;
+
+        } else {
+            throw new MissingTaskIndexException("Task Index Missing.");
+        }
+
+    }
+
+    /**
+     * Parses a user input string into a task.
+     *
+     * @param str The user input string.
+     * @return The parsed task.
+     * @throws InvalidDescriptionException If the task description is invalid.
+     * @throws InvalidDateTimeException    If the task date and time are invalid.
+     */
+    public static Task parseStringToTask(String str)
+            throws InvalidDescriptionException, InvalidDateTimeException {
+        Command command = getCommand(str);
+        switch(command) {
+        case TODO:
+            if (str.split(" ").length > 1) {
+                ToDo todo = new ToDo(str.split(" ")[1]);
+                return todo;
+            } else {
+                throw new InvalidDescriptionException("Invalid description.");
+            }
+        case DEADLINE:
+            if (str.split(" ").length > 3) {
+                String fullTaskDescription = str.split(" ", 2)[1];
+                String description = fullTaskDescription.split(" /by ")[0];
+                String by = fullTaskDescription.split(" /by ")[1];
+                String[] datetime = by.split(" ");
+                try {
+                    LocalDate date = LocalDate.parse(datetime[0]);
+                    LocalTime time = LocalTime.parse(datetime[1]);
+                    Deadline deadline = new Deadline(description, date, time);
+                    return deadline;
+                } catch (DateTimeParseException e) {
+                    throw new InvalidDateTimeException("Invalid Datetime.");
+                }
+            } else {
```

Comment:
Perhaps you could look into making the Happy Path prominent? swap out the if statement to test for the opposite, so that the unusual case is nested instead